src/plugins/usage_collection/server/report/store_application_usage.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import moment from 'moment';
+import { Writable } from '@kbn/utility-types';
+import { ISavedObjectsRepository } from 'src/core/server';
+import { ApplicationUsageReport } from './schema';
+import { getDailyId } from '../../common/application_usage';
+
+type WritableApplicationUsageReport = Writable<ApplicationUsageReport>;
+
+export const storeApplicationUsage = async (
+  repository: ISavedObjectsRepository,
+  appUsages: ApplicationUsageReport[],
+  timestamp: Date
+) => {
+  if (!appUsages.length) {
+    return;
+  }
+
+  const dayId = getDayId(timestamp);
+  const aggregatedReports = aggregateAppUsages(appUsages);
+
+  return Promise.allSettled(
+    aggregatedReports.map(async (report) => incrementUsageCounters(repository, report, dayId))
+  );
+};
+
+const aggregateAppUsages = (appUsages: ApplicationUsageReport[]) => {
+  return [
+    ...appUsages
+      .reduce((map, appUsage) => {
+        const key = getKey(appUsage);
+        const aggregated: WritableApplicationUsageReport = map.get(key) ?? {
+          appId: appUsage.appId,
+          viewId: appUsage.viewId,
+          minutesOnScreen: 0,
+          numberOfClicks: 0,
+        };
+
+        aggregated.minutesOnScreen += appUsage.minutesOnScreen;
+        aggregated.numberOfClicks += appUsage.numberOfClicks;
+
+        map.set(key, aggregated);
+        return map;
+      }, new Map<string, ApplicationUsageReport>())
+      .values(),
+  ];
+};
+
+const incrementUsageCounters = (
+  repository: ISavedObjectsRepository,
+  { appId, viewId, numberOfClicks, minutesOnScreen }: WritableApplicationUsageReport,
+  dayId: string
+) => {
+  const dailyId = getDailyId({ appId, viewId, dayId });
+
+  return repository.incrementCounter(
+    'application_usage_daily',
+    dailyId,
+    [
+      { fieldName: 'numberOfClicks', incrementBy: numberOfClicks },
+      { fieldName: 'minutesOnScreen', incrementBy: minutesOnScreen },
+    ],
+    {
+      upsertAttributes: {
+        appId,
+        viewId,
+        timestamp: getTimestamp(dayId),
+      },
```

Comment:
because of this comment: https://github.com/elastic/kibana/pull/94923#discussion_r597073842
```suggestion
      upsertAttributes: {
        appId,
        viewId,
        timestamp: getTimestamp(dayId),
        numberOfClicks,
        minutesOnScreen,
      },
```

Reply:
Yea, this is where we're going to be limited with TS support 😅 . 

The incremented fields (`counterFields`) are automatically added to the upsert attributes when the document is not present in the `incrementCounter` implementation.

So to fix what I said in https://github.com/elastic/kibana/pull/94923#discussion_r597073842:  the upsertAttributes needs to be all the mandatory attributes, except the attributes that are effectively incremented via the `counterFields` parameter 😅 ... 

However, that kind of constraint is too complex to express even in typescript (or it would just be unreadable), especially as `counterFields` can either be an array of string or an array of structured objects...